### ISSUE TITLE
Can’t join user to channel and to receive all messages of this channel.

### ISSUE THREAD

**Issue by polurezov8** [twilio/twilio-chat-ios#4]
Hi Twilio support team!
We have a project on the pre-release stage and face critical issue here.
### Description
We can’t join user to channel (type TCHChannel) and to receive all messages of this channel. 
So we see 2 problems here:
1) method channel.join returns false (result.isSuccesful)
2) method getLastMessgeWithCount - we don't receive a callback from it. When callback is not received - thus the rest of code is not executed.
### Steps to Reproduce
1. Login. I get  credentials from the backend. Credentials consist of token (String) and identity (String).
2. With a token, I execute the function to get chat client (type TwilioChatClient)
3. When I want to go to the chat dialogue screen, I perform the function of the one to connect the user to the channel. This function isn’t performed.
**Expected behavior:** successful connect to chat, getting all messages for this channel
**How Often Reproduces:** Randomly, in most cases. Sometimes I can connect to all chats and get messages, sometimes to one of all or none.
**Code for steps:**
**1.** Login:
```
    private func login() {
        chatChannelsManager.getTwilioToken(succeed: { [weak self] loadedCredentials in
            self?.credentials = loadedCredentials
```
Credentials model:
```
struct ChatCredentials {
    let token: String
    let identity: String
}
```
**2.** I execute the function to get chat client (type TwilioChatClient)
```
TwilioChatClient.chatClient(withToken: credentials.token, properties: nil, delegate: self) { [weak self] (result, chatClient) in
                guard let strongSelf = self else {
                    return
                }
                strongSelf.client = chatClient
                strongSelf.connectToAllChannels()
            }
```
```
private func connectToAllChannels() {
        for chat in activeChatList {
            connectToChannel(channelSid: chat.sid)
        }
    }
```
**Additional comment:** activeChatList - array with chats from my backend
After I’ll connect to channel by sid:
   ```
 private func connectToChannel(channelSid: String) {
        if let client = client, let channelsList = client.channelsList() {
            channelsList.channel(withSidOrUniqueName: channelSid, completion: { [weak self] (result, channel) in
                guard let strongSelf = self else {
                    return
                }
                if let channel = channel {
                    debugPrint("Added sid = \(String(describing: channel.sid))")
                    strongSelf.channels.insert(channel)
                    strongSelf.setUnreadMessageCount(channel: channel)
                }
            })
        }
    }
```
**3.** Join to chat when user want to start or continue chatting:
```
private func joinChat() {
        if let generalChannel = generalChannel {
            generalChannel.join(completion: { [weak self] result in
                guard let strongSelf = self else {
                    return
                }
                let canJoinChat = result.isSuccessful()
                debugPrint("joinChat status: \(canJoinChat)")
                debugPrint("generalChannel status: \(generalChannel.status.rawValue)")
                if let _ = generalChannel.messages {
                    DispatchQueue.main.async {
                        strongSelf.delegate?.didConnectToChat()
                    }
                    strongSelf.fetchChannelMessages(generalChannel)
                } else {
                    DispatchQueue.main.async {
                        strongSelf.delegate?.didFailConnectToChat()
                    }
                }
            })
        }
    }
```
**Additional comment:** canJoinChat = false, generalChannel status: 1 (TCHChannelStatusJoined,  ///< User is joined to the channel.)
### Logs
```
2018-11-20 12:34:43.728946+0200 Kego[6374:2787695] <<<<TwilioChatClient: 0x11bd08510>>>> TNTwilsockClientImpl[1]:        0x16df53000 | 11/20/12:34:43.728 | CRITICAL | 3      | TNTwilsockClientImpl | Message type list is empty
2018-11-20 12:34:43.729007+0200 Kego[6374:2787695] <<<<TwilioChatClient: 0x11bd08510>>>> TNRegNotificationCtxSession[2]:        0x16df53000 | 11/20/12:34:43.728 | WARNING  | 17     | TNRegNotificationCtxSession | Could not operate with notification context in twilsock. Error 16
2018-11-20 12:34:43.729044+0200 Kego[6374:2787695] <<<<TwilioChatClient: 0x11bd08510>>>> TNNotificationClient[0]:        0x16df53000 | 11/20/12:34:43.729 | FATAL    | 12     | TNNotificationClient | Fatal error while registering twilsock connection: 9: Message type list is empty, public error code: 0
2018-11-20 12:34:43.729127+0200 Kego[6374:2787717] <<<<TwilioChatClient: 0x11bd08510>>>> TNNotificationClient[2]:        0x16ea43000 | 11/20/12:34:43.729 | WARNING  | 12     | TNNotificationClient | onTwilsockError: 0 - Message type list is empty
2018-11-20 12:34:47.114080+0200 Kego[6374:2787809] <<<<TwilioChatClient: 0x11bd3a7e0>>>> TNTwilsockClientImpl[1]:        0x16f41b000 | 11/20/12:34:47.113 | CRITICAL | 45     | TNTwilsockClientImpl | Message type list is empty
2018-11-20 12:34:47.114140+0200 Kego[6374:2787809] <<<<TwilioChatClient: 0x11bd3a7e0>>>> TNRegNotificationCtxSession[2]:        0x16f41b000 | 11/20/12:34:47.114 | WARNING  | 56     | TNRegNotificationCtxSession | Could not operate with notification context in twilsock. Error 16
2018-11-20 12:34:47.114177+0200 Kego[6374:2787809] <<<<TwilioChatClient: 0x11bd3a7e0>>>> TNNotificationClient[0]:        0x16f41b000 | 11/20/12:34:47.114 | FATAL    | 51     | TNNotificationClient | Fatal error while registering twilsock connection: 9: Message type list is empty, public error code: 0
2018-11-20 12:34:47.114259+0200 Kego[6374:2787817] <<<<TwilioChatClient: 0x11bd3a7e0>>>> TNNotificationClient[2]:        0x16f5bf000 | 11/20/12:34:47.114 | WARNING  | 51     | TNNotificationClient | onTwilsockError: 0 - Message type list is empty
2018-11-20 12:34:48.176321+0200 Kego[6374:2787695] <<<<TwilioChatClient: 0x11bd3a7e0>>>> TNRegNotificationCtxSession[2]:        0x16df53000 | 11/20/12:34:48.176 | WARNING  | 17     | TNRegNotificationCtxSession | Could not operate with notification context in twilsock. Error 9
2018-11-20 12:34:48.176754+0200 Kego[6374:2787712] <<<<TwilioChatClient: 0x11bd3a7e0>>>> TNWebsocket[1]:        0x16e89f000 | 11/20/12:34:48.176 | CRITICAL | 4      | TNWebsocket | Received error while waiting for message: Exception: Bad socket descriptor
```
### Versions:
- Chat iOS SDK: TwilioChatClient 2.2.0
- Language: Swift 4
- Xcode: 10.1
- iOS Versions: 11.4.1; 12.0.1
- iOS Device: iPhone 6s, iPhone 7Plus, iPhone X, iPhone XS Max
Please review the problem and help to solve it. **This is very important for our project!**
Eagerly looking forward for your response! 
Thanks,

**Comment by ivantkachenko1riseapps** [comment]
@berkus , does information in logs, provided above by @polurezov8 , help to see where the problem is? 
Or some more info is needed from us?

**Comment by berkus** [comment]
No, full logs would be nice, just as I asked.

There's a [manual for android](https://github.com/twilio/twilio-chat-demo-android/blob/master/REPORT_BUGS.md) on capturing the logs and even though details are different, the general approach is the same - we need full logs to reconstruct context.

**Comment by rbeiter** [comment]
Hi @ivantkachenko1riseapps, for iOS please see https://www.twilio.com/docs/chat/error-handling-and-diagnostics#ios-logging-objective-c-swift for information on setting a higher log level to supply logs.  The logs may potentially contain PII so we encourage you send those via our [support website](https://www.twilio.com/console/support/tickets/create).

One additional note, in investigating this Channel's logs, we noticed an error indicating a member attempting to join that was already joined.  Joining a channel is only required once and beyond that first time will return an error.  I would recommend you check the `.status` property on the Channel object and ensure it is not `TCHChannelStatusJoined` before trying to join.

Please let us know if this helps!

Randy

**Comment by polurezov8** [comment]
@berkus I set the log level as you said, below is the log which are displayed in the console from the beginning of the application until the moment when we try to join the chat and receive messages.

Hi @rbeiter ! Thank you for your advice. I have added the execution logs below. I will try to change the joining chat for your advice, I’ll write you the result.

[LogForTwilioSupportTeam.txt](https://github.com/twilio/twilio-chat-ios/files/2605094/LogForTwilioSupportTeam.txt)

**Comment by ivantkachenko1riseapps** [comment]
Hi @berkus, 
How helpful are the logs provided by @polurezov8, is it clear now where exactly the problem is?

**Comment by polurezov8** [comment]
Hi @berkus and @rbeiter! I tried to change the logic of working with chats, the results are described below:

**1.** I have a table with all the chat rooms.

**2.** When a user clicks on a cell (choses some chat), I open chat dialog screen
The method joinToChatAndFetchMessages() finds the chat by sid:

joinAndFetchMessagesWith() method:

```
public func joinAndFetchMessagesWith(_ chat: Chat) {
        if let channel = channels.first(where: { $0.sid == chat.sid }) {
            generalChannel = channel
            joinChat()
        } else {
            DispatchQueue.main.async {
                self.delegate?.didFailConnectToChat()
            }
        }
    }
```

**3.** Once the necessary chat is found I’m joining to it with method join()
In the method join(), I check the status of the channel whether it's equal to 0 (zero)

I can be NOT equal to 0 in such cases:
- a) TCHChannelStatusJoined or 1 - User is joined to the channel.
- b) TCHChannelStatusNotParticipating or 2 - User is not participating on this channel.
- c) TCHChannelStatusUnknown or 3 - User’s status on this channel is not known, used on instances of TCHChannelDescriptor.

joinChat() method:

```
private func joinChat() {
        if let generalChannel = generalChannel {
            if generalChannel.status.rawValue == 0 { // 0 - User is invited to the channel but not joined.
                generalChannel.join(completion: { result in
                    let canJoinChat = result.isSuccessful()
                    debugPrint("""
                        JoinChat status: \(canJoinChat)
                        GeneralChannel status: \(generalChannel.status.rawValue)
                        """)
                })
            }
            if let _ = generalChannel.messages {
                DispatchQueue.main.async {
                    self.delegate?.didConnectToChat()
                }
                self.fetchChannelMessages(generalChannel)
            } else {
                DispatchQueue.main.async {
                    self.delegate?.didFailConnectToChat()
                }
            }
        }
    }
```

**3.1** Channels that I can not get messages from - their status == 1 // User is joined to the channel.
But for some channels with status == 1 I can get its messages.
If the status is NOT 0, I’m trying to get messages with method fetchChannelMessages()

fetchChannelMessages() method:

```
private func fetchChannelMessages(_ channel: TCHChannel) {
        channel.getMessagesCount { (_, count) in
            let messagesCount = count
            debugPrint("messagesCount = \(messagesCount)")
            if messagesCount > 0 {
                channel.messages?.getLastWithCount(messagesCount, completion: { [weak self]  (result, messages) in
                    guard let strongSelf = self else {
                        return
                    }
                    if result.isSuccessful() {
                        if let messages = messages {
                            strongSelf.messages = messages
                        }
                    }
                    DispatchQueue.main.async {
                        strongSelf.delegate?.didFetchMessagesHistory(strongSelf.messages)
                    }
                })
                channel.messages?.setAllMessagesConsumedWithCompletion { [weak self] (result, count) in
                    guard let strongSelf = self else {
                        return
                    }
                    strongSelf.setUnreadMessageCount(channel: channel)
                }
            } else {
                self.delegate?.didEndConnectToChat()
            }
        }
    }
```

But I can’t get messages from this channel, because generalChannel.messages == nil.

I can not understand why. Channel statuses are the same (joined), but for some chat rooms I cannot receive messages. Guys, please **HELP!**

**Comment by berkus** [comment]
You must wait for channel synchronization status to become `.ALL` before querying messages. I believe this is noted in the documentation.

**Comment by berkus** [comment]
See https://media.twiliocdn.com/sdk/ios/chat/releases/2.3.3/docs/Constants/TCHChannelSynchronizationStatus.html

**Comment by berkus** [comment]
See https://media.twiliocdn.com/sdk/android/chat/releases/3.1.1/docs/enumcom_1_1twilio_1_1chat_1_1Channel_1_1SynchronizationStatus.html#details

**Comment by polurezov8** [comment]
@berkus @rbeiter I checked channels synchronization statuses. 
The status of the channels to which I can connect and can not connect at the moment of selecting a chat is 3 // TCHChannelSynchronizationStatusAll
Most likely the problem is not synchronization status.
I still get nil when I try to receive a message and doing optional binding:

```
if let _ = channel.messages { // nil
...
}
```

**Comment by ivantkachenko1riseapps** [comment]
Hi @berkus , @rbeiter , 
is it clear from @polurezov8 's comment above what's the key of the problem and how we can solve it? 
Is any more information should be provided from our side?

**Comment by ivantkachenko1riseapps** [comment]
Hello, @berkus , @rbeiter, 
Hope you had good weekend :)
Please let us know what you think about it, what should we do to fix this issue.

**Comment by ivantkachenko1riseapps** [comment]
Morning @berkus, @rbeiter ,
when will you be able to provide some response?
I want the emphasize that this issue fixing is **CRITICAL** for our project success. And currently it's the only one that prevents us from the release, we need to handle it **ASAP!** We've chosen Twilio to as a an advanced chat solution but currently it became a **project stopper**..
We are eagerly waiting for your help!

**Comment by berkus** [comment]
Hi,

regarding this code

```
private func joinChat() {
            ....
            if let _ = generalChannel.messages {
                ....
                self.fetchChannelMessages(generalChannel)
            }
```

Is `fetchChannelMessages()` called?

**Comment by berkus** [comment]
Also, what I can see from the logs is that TwilioChatClient gets deinitializer prior to your call to `channel.join()` so it seems that the client is already destructed by the time you try to join and therefore fails.

Could you check how you hold a reference to the chat client in your code?

**Comment by polurezov8** [comment]
Hi @berkus! The above code called when generalChannel.massages not equal nil. In the case when generalChannel.messages == nil method fetchChannelMessages() not called.

Yes, of course, I will check and write you the result.

**Comment by berkus** [comment]
The logs show this (these are sequence of the calls by the SDK):
```
11/21/18:29:48.094 | INFO     | IPM | client [api] shutdown <-- note shutdown here!
11/21/18:29:48.100 | INFO     | IPM | client [api] get channels
11/21/18:29:48.101 | INFO     | IPM | client [api] get users
11/21/18:29:48.101 | INFO     | IPM | channels: [api] get user channels
11/21/18:29:48.172 | INFO     | IPM | channels: [api] get channel CHd64854c4696b4d69bc62dc2ffd2e5463
....
11/21/18:29:48.195 | INFO     | IPM | channels: [api] get channel CHb4b8d2f8947b4b0c9ee39befefd63829
11/21/18:29:52.112 | INFO     | IPM | [M31Ru] channel: [api] join <-- note join here!
11/21/18:29:52.189 | INFO     | IPM | [M31Ru] channel: [api] get messages
```

**Comment by polurezov8** [comment]
@berkus Interesting, I never call method client.shutdown() and when I set breakpoint on - (void)shutdown; in TwilioChatClient.h I never stop at it when my program runs. 
Can other functions implicitly call shutdown() method?

**Comment by berkus** [comment]
This could happen if reference to ChatClient is dropped. Where do you keep it?

**Comment by polurezov8** [comment]
About reference:
After I got the token from my backend and run the method:

```
            TwilioChatClient.chatClient(withToken: credentials.token, properties: nil, delegate: self) { [weak self] (result, chatClient) in
                guard let strongSelf = self else {
                    return
                }
                strongSelf.client = chatClient
            }
```

And I use a strong reference for object "client".
Variable client is stored in ChatManager class:
`var client: TwilioChatClient?`
In the same class all the methods associated with working with chats.

**Comment by berkus** [comment]
Yeah, this should keep the reference alive then.

**Comment by berkus** [comment]
Could you add some logging in your workflow to detect at which point in time relative to your code the ChatClient.shutdown gets called? Maybe this gives an insight.

**Comment by polurezov8** [comment]
Yes, sure. I'll add debugPrint to all chat's methods and write logs here.

**Comment by berkus** [comment]
I don't see any explicit calls to shutdown, but I do see that you seem to have two instances of TwilioChatClient operating - not sure if this is a bug or supposed to be like this.

**Comment by polurezov8** [comment]
Hi @berkus @rbeiter!
I analyzed and refactored existing code. To fix the problem we wrote about, we had to do:
**1**. Check variable client and only if it is equal to nil to request access from my backend. Previously I accessed every time I logged in to the app. Thereby overwriting the variable client (because of this, the method shutdown() was called).
```
private func login() {
     if client == nil {
          chatChannelsManager.getTwilioToken(succeed: { [weak self] loadedCredentials in
          ...
```
**2**. I implemented TwilioAccessManager. Special thanks for such a handy tool 👍🏻. 
Thank you very much for your help in solving this problem! Have a nice day!
